### PR TITLE
This implements "Provide ability to configure whether a new iteration is a new user or same user" (BUG 62861)

### PR DESCRIPTION
**Pull request by pmouawad**
## Description

UbikLoadPack team is happy to contribute this PR that provides ability to configure whether a new iteration is a new user or same user

Regards
Qian Qian and Philippe
https://ubikloadpack.com
@ubikloadpack

## Motivation and Context

https://bz.apache.org/bugzilla/show_bug.cgi?id=62861

## How Has This Been Tested?

JUnit tests and manual testing

## Types of changes
- New feature (non-breaking change which adds functionality)

## Checklist:

- [X] My code follows the [code style][style-guide] of this project.
- [X] I have updated the documentation accordingly.

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

### CODE REVIEW

File: src/core/org/apache/jmeter/threads/JMeterThread.java [hunk 1]
```diff
@@ -148,7 +150,7 @@
 
     private final ReentrantLock interruptLock = new ReentrantLock(); // ensure that interrupt cannot overlap with shutdown
 
-    public JMeterThread(HashTree test, JMeterThreadMonitor monitor, ListenerNotifier note) {
+    public JMeterThread(HashTree test, JMeterThreadMonitor monitor, ListenerNotifier note,Boolean isSameUser) {
```
Comment (vlsi) [inline src/core/org/apache/jmeter/threads/JMeterThread.java]: Parameter naming is kind of hard to understand.
If you add a parameter to existing constructor/method, you might want to just add another constructor and delegate. That would prevent adding `,false)` all over the place

**Review by vlsi — CHANGES_REQUESTED** [review]
Did you consider making "reset thread iteration" a property of a LoopController rather than a thread?

I looks like loop is more relevant component to "reset the state on each iteration"

File: src/core/org/apache/jmeter/threads/JMeterThread.java (hunk 1 above)
Comment (ubikloadpack) [reply src/core/org/apache/jmeter/threads/JMeterThread.java]: Agreed

**Comment by ubikloadpack** [comment]
> Did you consider making "reset thread iteration" a property of a LoopController rather than a thread?
> 
> I looks like loop is more relevant component to "reset the state on each iteration"

I don't think so, as it's more about Thread (Virtual User) configuration, ie what kind of user they simulate.
Not sure what meaning it would have applied to LoopController

**Comment by vlsi** [comment]
> I don't think so, as it's more about Thread (Virtual User) configuration, ie what kind of user they simulate.

What about custom thread groups?

**Comment by pmouawad** [comment]
> > I don't think so, as it's more about Thread (Virtual User) configuration, ie what kind of user they simulate.
> 
> What about custom thread groups?
@vlsi 
Do you propose to  move setIsSameUser/isSameUser into AbstractThreadGroup ?

**Comment by vlsi** [comment]
> Do you propose to move setIsSameUser/isSameUser into AbstractThreadGroup ?

As for me, "new iteration is a new user / same user" does not belong to a Thread. It belongs to whatever is controlling the iterations. And it turns out JMeterThread is reusing loop parameters, so it looks like "reset user" should be there.

What if we would like to add "relogin user after each 15 min" and so on?
That would bloat JMeterThread with irrelevant features, especially when load protocol has no notion of "user".
For instance, if JMeter is used to drive TCP load, then "new iteration is a new user" would be obscure.


Could the same behavior be configured by adding a listener/configuration element?
If that is possible, why don't we go with that approach?

File: src/core/org/apache/jmeter/threads/JMeterThread.java [hunk 2]
```diff
@@ -95,7 +95,7 @@
     private final TestCompiler compiler;
 
     private final JMeterThreadMonitor monitor;
-
+
```
Comment (vlsi) [inline src/core/org/apache/jmeter/threads/JMeterThread.java]: Please avoid adding trailing whitespace

File: src/core/org/apache/jmeter/threads/JMeterThread.java [hunk 3]
```diff
@@ -123,8 +123,10 @@
 
     private long endTime = 0;
 
-    private boolean scheduler = false;
+    private boolean isSameUserOnNextIteration = false;
```
Comment (vlsi) [inline src/core/org/apache/jmeter/threads/JMeterThread.java]: Is this `final` or not?
If the variable is not final, when what sense does it make to add a constructor argument for it?

File: src/core/org/apache/jmeter/threads/ThreadGroup.java [hunk 4]
```diff
@@ -621,6 +622,7 @@ public void run() {
             } catch (Exception ex) {
                 log.error("An error occurred scheduling delay start of threads for Thread Group: {}", getName(), ex);
             }
+
```
Comment (vlsi) [inline src/core/org/apache/jmeter/threads/ThreadGroup.java]: was this intended?

File: src/core/org/apache/jmeter/threads/gui/ThreadGroupGui.java [hunk 5]
```diff
@@ -228,14 +244,15 @@ private void init() { // WARNING: called from ctor so must not be overridden (i.
...
-
+        threadPropsPanel.add(createUserOptionsPanel());  
         if (showDelayedStart) {
             delayedStart = new JCheckBox(JMeterUtils.getResString("delayed_start")); // $NON-NLS-1$
             threadPropsPanel.add(delayedStart);
         }
         scheduler = new JCheckBox(JMeterUtils.getResString("scheduler")); // $NON-NLS-1$
         scheduler.addItemListener(this);
         threadPropsPanel.add(scheduler);
+
```
Comment (vlsi) [inline src/core/org/apache/jmeter/threads/gui/ThreadGroupGui.java]: was this intended?

File: src/protocol/http/org/apache/jmeter/protocol/http/control/CookieManager.java [hunk 6]
```diff
@@ -428,7 +438,9 @@ public void testEnded(String host) {
     /** {@inheritDoc} */
     @Override
     public void testIterationStart(LoopIterationEvent event) {
-        if (getClearEachIteration()) {
+        JMeterVariables jMeterVariables = JMeterContextService.getContext().getVariables();
+        if ((getControlledByThread() && !jMeterVariables.isSameUserOnNextIteration()) 
+                || getClearEachIteration()) {
```
Comment (vlsi) [inline src/protocol/http/org/apache/jmeter/protocol/http/control/CookieManager.java]: Does that mean `getClearEachIteration` impacts the result no matter of `getControlledByThread`?

Comment (pmouawad) [reply src/protocol/http/org/apache/jmeter/protocol/http/control/CookieManager.java]: Yes, but if you look at GUI it's explicit

**Comment by pmouawad** [comment]
The following changes will be made on PR:

- Replace in GUIs the 2 checkboxes Clear XXX / Use Thread Group by a Radio box to make it more clear they are exlusive
- Select by default in CacheManager, CookieManager and Auth Manager "Use Thread Group..."
- Documentation will be updated

**Comment by pmouawad** [comment]
> > Do you propose to move setIsSameUser/isSameUser into AbstractThreadGroup ?
> 
> As for me, "new iteration is a new user / same user" does not belong to a Thread. It belongs to whatever is controlling the iterations. And it turns out JMeterThread is reusing loop parameters, so it looks like "reset user" should be there.

I don’ share your opinion :
Loop Controller is more of an algorithmic element and has no notion of user.
Thread Group control user behaviour:
- how many
- how they ramp up
- how much they run

And now we’re adding what kind are they.
Another new element makes it maybe more flexible but more complex.



> 
> What if we would like to add "relogin user after each 15 min" and so on?

how would it be impacted by this feature?
- can you illustrate how you would do it now and how this feature impacts it ?


> That would bloat JMeterThread with irrelevant features, especially when load protocol has no notion of "user".
> For instance, if JMeter is used to drive TCP load, then "new iteration is a new user" would be obscure.

so it would just be ignored 
> 
> Could the same behavior be configured by adding a listener/configuration element?
> If that is possible, why don't we go with that approach?
Feel free to propose a PR illustrating it or amending this one.

File: src/protocol/http/org/apache/jmeter/protocol/http/control/AuthManager.java [hunk 7]
```diff
@@ -555,8 +555,8 @@ public void testEnded(String host) {
     @Override
     public void testIterationStart(LoopIterationEvent event) {
         JMeterVariables jMeterVariables = JMeterContextService.getContext().getVariables();
-        if ((getControlledByThread() && !jMeterVariables.isSameUserOnNextIteration()) 
-                || getClearEachIteration()) {
+        if ((getControlledByThread() && !jMeterVariables.isSameUserOnNextIteration())
+                || (!getControlledByThread() && getClearEachIteration())) {
```
Comment (vlsi) [inline src/protocol/http/org/apache/jmeter/protocol/http/control/AuthManager.java]: Would ternary help here?

**Comment by vlsi** [comment]
Please compress pngs

**Pull request merged**